### PR TITLE
feat: ascatNgs PDF delivery 

### PR DESCRIPTION
**Pull request by ivadym**
### This PR:

Adds a new delivery file in BALSAMIC(`PDF`) containing the main `ascatNgs` output plots as well as a table with some sample statistics.

#### Steps

- [x] Merge the `.png` plots and the statistics `.txt` file into a single `PDF`.
- [x] Create a new snakemake rule to save the generated PDF.

### Review and tests: 
- [ ] Tests pass
- [ ] Code review
- [ ] New code is executed and covered by tests, and test approve

### CODE REVIEW

**Comment by ivadym** [comment]
Currently generated PDF: 

[ascat.output.pdf](https://github.com/Clinical-Genomics/BALSAMIC/files/7627167/ascat.output.pdf)

@ashwini06 @khurrammaqbool

File: BALSAMIC/snakemake_rules/variant_calling/somatic_sv_tumor_normal.rule [hunk 1]
```diff
@@ -165,25 +164,54 @@ ascat.pl \
...
 
-tabix -p vcf -f {output.finalvcf};
+tabix -p vcf -f {output.final_vcf};
 
 echo -e \"{params.tumor}\\tTUMOR\\n{params.normal}\\tNORMAL\" > {output.namemap};
 
 rm -rf {params.tmpdir};
         """
+
+rule ascat_tumor_normal_merge_output:
```
Comment (hassanfa) [inline BALSAMIC/snakemake_rules/variant_calling/somatic_sv_tumor_normal.rule:190]: This rule is going to be submitted by default value of resources. Although it'll take a very short time. Either add it as part of `rule all` and tag it separately, or add rule name to https://github.com/Clinical-Genomics/BALSAMIC/blob/v8.2.3/BALSAMIC/config/cluster.json

File: BALSAMIC/snakemake_rules/variant_calling/somatic_sv_tumor_normal.rule [hunk 2]
```diff
@@ -165,25 +164,54 @@ ascat.pl \
...
+    input:
+        sample_statistics = vcf_dir + "CNV.somatic." + config["analysis"]["case_id"] + ".ascat.samplestatistics.txt",
+        ascat_profile_plot= vcf_dir + "CNV.somatic." + config["analysis"]["case_id"] + ".ascat.ascatprofile.png",
+        raw_profile_plot= vcf_dir + "CNV.somatic." + config["analysis"]["case_id"] + ".ascat.rawprofile.png",
+        aspcf_plot = vcf_dir + "CNV.somatic." + config["analysis"]["case_id"] + ".ascat.ASPCF.png",
+        tumor_plot = vcf_dir + "CNV.somatic." + config["analysis"]["case_id"] + ".ascat.tumor.png",
+        germline_plot = vcf_dir + "CNV.somatic." + config["analysis"]["case_id"] + ".ascat.germline.png",
+        sunrise_plot = vcf_dir + "CNV.somatic." + config["analysis"]["case_id"] + ".ascat.sunrise.png"
```
Comment (hassanfa) [inline BALSAMIC/snakemake_rules/variant_calling/somatic_sv_tumor_normal.rule:198]: You should consider crunching all these into one:
```suggestion
        sample_statistics = vcf_dir + "CNV.somatic." + config["analysis"]["case_id"] + ".ascat.samplestatistics.txt",
        ascat_plots = expand(vcf_dir + "CNV.somatic." + config["analysis"]["case_id"] + "{output_suffix}", [ ".ascat.ascatprofile.png",".ascat.rawprofile.png" ...])`
```

File: BALSAMIC/snakemake_rules/variant_calling/somatic_sv_tumor_normal.rule [hunk 3]
```diff
@@ -165,25 +164,54 @@ ascat.pl \
...
+            output.ascat_output_pdf,
+            input.sample_statistics,
+            [
+                input.ascat_profile_plot,
+                input.raw_profile_plot,
+                input.aspcf_plot,
+                input.tumor_plot,
+                input.germline_plot,
+                input.sunrise_plot
+            ]
```
Comment (hassanfa) [inline BALSAMIC/snakemake_rules/variant_calling/somatic_sv_tumor_normal.rule:216]:
```suggestion
input.ascat_plots
```

File: BALSAMIC/snakemake_rules/variant_calling/somatic_sv_tumor_normal.rule [hunk 4]
```diff
@@ -165,25 +164,54 @@ ascat.pl \
...
+    input:
+        sample_statistics = vcf_dir + "CNV.somatic." + config["analysis"]["case_id"] + ".ascat.samplestatistics.txt",
+        ascat_profile_plot= vcf_dir + "CNV.somatic." + config["analysis"]["case_id"] + ".ascat.ascatprofile.png",
+        raw_profile_plot= vcf_dir + "CNV.somatic." + config["analysis"]["case_id"] + ".ascat.rawprofile.png",
+        aspcf_plot = vcf_dir + "CNV.somatic." + config["analysis"]["case_id"] + ".ascat.ASPCF.png",
+        tumor_plot = vcf_dir + "CNV.somatic." + config["analysis"]["case_id"] + ".ascat.tumor.png",
+        germline_plot = vcf_dir + "CNV.somatic." + config["analysis"]["case_id"] + ".ascat.germline.png",
+        sunrise_plot = vcf_dir + "CNV.somatic." + config["analysis"]["case_id"] + ".ascat.sunrise.png"
+    output:
+        ascat_output_pdf = vcf_dir + "CNV.somatic." + config["analysis"]["case_id"] + ".ascat.output.pdf"
```
Comment (hassanfa) [inline BALSAMIC/snakemake_rules/variant_calling/somatic_sv_tumor_normal.rule:200]: In this output properly part of https://github.com/Clinical-Genomics/BALSAMIC/blob/595caae47f5198620c963f07275c95ba914b56d0/BALSAMIC/workflows/balsamic.smk#L317 ?

File: BALSAMIC/snakemake_rules/variant_calling/somatic_sv_tumor_normal.rule [hunk 5]
```diff
@@ -165,25 +164,54 @@ ascat.pl \
...
+    params:
+        housekeeper_id = {"id": config["analysis"]["case_id"], "tags": "research"}
+    run:
+        from BALSAMIC.utils.workflowscripts import save_ascat_output_pdf
+
+        save_ascat_output_pdf(
+            output.ascat_output_pdf,
```
Comment (hassanfa) [inline BALSAMIC/snakemake_rules/variant_calling/somatic_sv_tumor_normal.rule:207]: Possible to make this an standalone script have it as `shell` so we can run it within BALSAMIC container instead?

**Review by hassanfa — COMMENTED** [review]
Some comments from my side.

File: BALSAMIC/snakemake_rules/variant_calling/somatic_sv_tumor_normal.rule (hunk 2 above)
Comment (ashwini06) [reply BALSAMIC/snakemake_rules/variant_calling/somatic_sv_tumor_normal.rule:198]: even more generalization..
`ascat_plots = expand(vcf_dir + "CNV.somatic." + config["analysis"]["case_id"] + ".ascat." + "{output_suffix}" + ".png",  output_suffix=[ "ascatprofile", "rawprofile", "ASPCF", ...])`

File: BALSAMIC/snakemake_rules/variant_calling/somatic_sv_tumor_normal.rule (hunk 4 above)
Comment (ashwini06) [reply BALSAMIC/snakemake_rules/variant_calling/somatic_sv_tumor_normal.rule:200]: @ivadym : you should add this output file to analysis_specific_results in balsamic.smk

**Review by ashwini06 — COMMENTED** [review]
Nice implementation @ivadym 👍🏼

File: BALSAMIC/snakemake_rules/variant_calling/somatic_sv_tumor_normal.rule (hunk 5 above)
Comment (ivadym) [reply BALSAMIC/snakemake_rules/variant_calling/somatic_sv_tumor_normal.rule:207]: If you mean to run it inside `/balsamic_cache/X.X.X/containers/balsamic.sif`, I think we can address it on another PR (currently it is not downloaded when the balsamic cache is generated. I believe that we have not included it within the VALID_CONTAINER_CONDA_NAME constant).

File: BALSAMIC/snakemake_rules/variant_calling/somatic_sv_tumor_normal.rule [hunk 6]
```diff
@@ -165,25 +162,49 @@ ascat.pl \
...
+cp {params.tmpdir}/{params.tumor}.sunrise.png {output.ascat_plots[2]};
 
-cp {params.tmpdir}/{params.tumor}.tumour.png {output.tumorplot};
+cp {params.tmpdir}/{params.tumor}.tumour.png {output.ascat_plots[3]};
 
-cp {params.tmpdir}/{params.tumor}.ASCATprofile.png {output.ascatprofileplot};
+cp {params.tmpdir}/{params.tumor}.ASCATprofile.png {output.ascat_plots[4]};
 
-cp {params.tmpdir}/{params.tumor}.ASPCF.png {output.aspcfplot};
+cp {params.tmpdir}/{params.tumor}.ASPCF.png {output.ascat_plots[5]};
```
Comment (ivadym) [inline BALSAMIC/snakemake_rules/variant_calling/somatic_sv_tumor_normal.rule:179]: I have tried to generalise this part as well but I am not sure how to do it as the path is determined from the parameters and not from wildcards.

**Review by ivadym — COMMENTED** [review]
Thank you for your comments and time @hassanfa and @ashwini06. I followed your suggestions but feel free to add any comments you consider appropriate.

Job rules that are run in balsamic dry-mode @ashwini06:

Screen Shot 2021-12-05 at 23 53 22

Output PDF:

[CNV.somatic.TN_WGS.ascat.output.pdf](https://github.com/Clinical-Genomics/BALSAMIC/files/7659367/CNV.somatic.TN_WGS.ascat.output.pdf)

File: BALSAMIC/snakemake_rules/variant_calling/somatic_sv_tumor_normal.rule [hunk 7]
```diff
@@ -165,25 +162,49 @@ ascat.pl \
...
   -normal {input.bamN};
 
-cp  {params.tmpdir}/{params.tumor}.copynumber.caveman.vcf.gz {output.finalvcf};
+cp  {params.tmpdir}/{params.tumor}.copynumber.caveman.vcf.gz {output.final_vcf};
 
-cp {params.tmpdir}/{params.tumor}.samplestatistics.txt {output.samplestatistics};
+cp {params.tmpdir}/{params.tumor}.samplestatistics.txt {output.sample_statistics};
 
-cp {params.tmpdir}/{params.tumor}.germline.png {output.germlineplot};
+cp {params.tmpdir}/{params.tumor}.germline.png {output.ascat_plots[0]};
```
Comment (ashwini06) [inline BALSAMIC/snakemake_rules/variant_calling/somatic_sv_tumor_normal.rule]:
```suggestion
cp {params.tmpdir}/{params.tumor}.germline.png {output.ascat_plots[4]};
```

File: BALSAMIC/snakemake_rules/variant_calling/somatic_sv_tumor_normal.rule [hunk 8]
```diff
@@ -165,25 +162,49 @@ ascat.pl \
...
 
-cp {params.tmpdir}/{params.tumor}.rawprofile.png {output.rawprofileplot};
+cp {params.tmpdir}/{params.tumor}.rawprofile.png {output.ascat_plots[1]};
 
-cp {params.tmpdir}/{params.tumor}.sunrise.png {output.sunriseplot};
+cp {params.tmpdir}/{params.tumor}.sunrise.png {output.ascat_plots[2]};
```
Comment (ashwini06) [inline BALSAMIC/snakemake_rules/variant_calling/somatic_sv_tumor_normal.rule]:
```suggestion
cp {params.tmpdir}/{params.tumor}.sunrise.png {output.ascat_plots[5]};
```

File: BALSAMIC/snakemake_rules/variant_calling/somatic_sv_tumor_normal.rule [hunk 9]
```diff
@@ -165,25 +162,49 @@ ascat.pl \
...
 
-cp {params.tmpdir}/{params.tumor}.tumour.png {output.tumorplot};
+cp {params.tmpdir}/{params.tumor}.tumour.png {output.ascat_plots[3]};
 
-cp {params.tmpdir}/{params.tumor}.ASCATprofile.png {output.ascatprofileplot};
+cp {params.tmpdir}/{params.tumor}.ASCATprofile.png {output.ascat_plots[4]};
```
Comment (ashwini06) [inline BALSAMIC/snakemake_rules/variant_calling/somatic_sv_tumor_normal.rule]:
```suggestion
cp {params.tmpdir}/{params.tumor}.ASCATprofile.png {output.ascat_plots[0]};
```

File: BALSAMIC/snakemake_rules/variant_calling/somatic_sv_tumor_normal.rule (hunk 6 above)
Comment (ashwini06) [inline BALSAMIC/snakemake_rules/variant_calling/somatic_sv_tumor_normal.rule]:
```suggestion
cp {params.tmpdir}/{params.tumor}.ASPCF.png {output.ascat_plots[2]};
```

Comment (ashwini06) [reply BALSAMIC/snakemake_rules/variant_calling/somatic_sv_tumor_normal.rule]: Capturing the output files from the list is fine. But just make sure you have the correct order of filenames

**Review by ashwini06 — CHANGES_REQUESTED** [review]

File: BALSAMIC/snakemake_rules/variant_calling/somatic_sv_tumor_normal.rule (hunk 7 above)
Comment (ivadym) [reply BALSAMIC/snakemake_rules/variant_calling/somatic_sv_tumor_normal.rule]: I updated this in a previous PR (6aca948) this morning.

File: BALSAMIC/snakemake_rules/variant_calling/somatic_sv_tumor_normal.rule (hunk 6 above)
Comment (ivadym) [reply BALSAMIC/snakemake_rules/variant_calling/somatic_sv_tumor_normal.rule]: I updated this in a previous PR (6aca948) this morning.

Comment (hassanfa) [reply BALSAMIC/snakemake_rules/variant_calling/somatic_sv_tumor_normal.rule]: So all the PNG files from tmpdir are copied to output, right? Will a solution like this work? `cp {params.tmp_dir}/*.png {params.output_dir}`

File: BALSAMIC/snakemake_rules/variant_calling/somatic_sv_tumor_normal.rule (hunk 5 above)
Comment (hassanfa) [reply BALSAMIC/snakemake_rules/variant_calling/somatic_sv_tumor_normal.rule:207]: fair enough 👍🏽

File: BALSAMIC/snakemake_rules/variant_calling/somatic_sv_tumor_normal.rule (hunk 6 above)
Comment (hassanfa) [reply BALSAMIC/snakemake_rules/variant_calling/somatic_sv_tumor_normal.rule]: True. All input are TUMOR.XXXX.png. So the string `TUMOR` needs to be replaced by `CNV.somatic.{params.case_name}.ascat`. For future reference something like would work:

```bash
find {params.tmpdir} -name "*.png" \
| while read filename; \
do new_name=$(echo "${{CNV.somatic.{params.case_name}.ascat/{params.tumor}/"$secondString"}}");\
cp "$filename" "$new_name";
done
```

Comment (hassanfa) [reply BALSAMIC/snakemake_rules/variant_calling/somatic_sv_tumor_normal.rule]: It's ugly though 😛

Comment (ivadym) [reply BALSAMIC/snakemake_rules/variant_calling/somatic_sv_tumor_normal.rule]: Beautiful in it's own way*

Maybe it's better to have a snakemake based solution for this, though I couldn’t find anything

Comment (khurrammaqbool) [reply BALSAMIC/snakemake_rules/variant_calling/somatic_sv_tumor_normal.rule]: We can skip the png outputs and report/deliver only one pdf as it includes all the information, unless we would like to parse the text file for scout etc.

**Review by ashwini06 — APPROVED** [review]

File: BALSAMIC/snakemake_rules/variant_calling/somatic_sv_tumor_normal.rule [hunk 10]
```diff
@@ -195,16 +195,13 @@ rule ascat_tumor_normal_merge_output:
     output:
         ascat_output_pdf = vcf_dir + "CNV.somatic." + config["analysis"]["case_id"] + ".ascat.output.pdf"
     params:
-        housekeeper_id = {"id": config["analysis"]["case_id"], "tags": "research"}
+        housekeeper_id = {"id": config["analysis"]["case_id"], "tags": "research"},
+        merge_ascat_output_script= get_script_path("merge_ascat_output.py"),
```
[the remark this answers is not in the document]
Comment (ivadym) [reply BALSAMIC/snakemake_rules/variant_calling/somatic_sv_tumor_normal.rule:199]: I agree 😄

**Review by ashwini06 — APPROVED** [review]
💯

**Pull request merged**